### PR TITLE
Add possible implementation of aligning Z, 1-qubit, and 2-qubit gates optimizer

### PR DESCRIPTION
**Pull request by KevinVillela**
This is a bit of a strawman implementation of the optimizer that separates moments into Z ops, other 1-qubit ops, and 2-qubit ops. I tried for a while to find a good straightforward algorithm to do this, but I almost immediately ran into a situation where said algorithm would break down (see test files for examples).

On the other hand, I found that using Simulated Annealing resulted in very good results that only took a few seconds to compute (and this could be sped up drastically). I know that it seems like overkill for what we want, so I thought I'd send a PR so we can discuss it here. Thanks!

### CODE REVIEW

File: cirq/google/optimizers/align_interactions_test.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
+from typing import Callable, List, Dict
```
Comment (vtomole) [inline cirq/google/optimizers/align_interactions_test.py]: Add copyright

File: cirq/google/optimizers/align_interactions.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
+import random
```
Comment (vtomole) [inline cirq/google/optimizers/align_interactions.py]: Add copyright

**Comment by mpharrigan** [comment]
Why is it so hard? Does the problem become simpler if you're assured that someone has run the single qubit merger optimizer first? Right now that merges all adjacent single qubit rotations into at most one PhX and Z. As a side issue; we should make that optimizer output the new PhXZ gate.

File: cirq/google/optimizers/align_interactions_test.py (hunk 1 above)
Comment (KevinVillela) [reply cirq/google/optimizers/align_interactions_test.py]: Done.

File: cirq/google/optimizers/align_interactions.py (hunk 2 above)
Comment (KevinVillela) [reply cirq/google/optimizers/align_interactions.py]: Done.

**Comment by KevinVillela** [comment]
> Why is it so hard? Does the problem become simpler if you're assured that someone has run the single qubit merger optimizer first? Right now that merges all adjacent single qubit rotations into at most one PhX and Z. As a side issue; we should make that optimizer output the new PhXZ gate.

I don't think so. Assuming there are only two types of gates, X and SWAP, and X's can't go past SWAPs, then I think the problem is still a CSP, and the difficulty lies in finding the minimum number of moments. Take the following example:
```
              ┌───────────┐               ┌──────────┐
0: ────────────iSwap──────────X───────X────iSwap─────────────X───────
               │                           │
1: ───iSwap────iSwap──────────X───────X────┼────iSwap────────iSwap───
      │                                    │    │            │
2: ───iSwap────iSwap──────────X───────X────┼────iSwap────────iSwap───
               │                           │
3: ────────────┼────iSwap─────────────X────iSwap─────────────────────
               │    │
4: ───X────────┼────┼────X────X───────X────iSwap─────────────X───────
               │    │                      │
5: ────────────┼────iSwap─────X────────────┼─────────────────X───────
               │                           │
6: ───iSwap────┼────X─────────────X────────┼─────────────────iSwap───
      │        │                           │                 │
7: ───iSwap────iSwap───────────────────────iSwap─────────────iSwap───
              └───────────┘               └──────────┘
```
A solution is 
```
              ┌──────────┐                           ┌──────────┐
0: ────────────iSwap─────────X───────X─────────────────────iSwap────X───
               │                                           │
1: ───iSwap────iSwap─────────X───X───────────iSwap────iSwap┼────────────
      │                                      │        │    │
2: ───iSwap────iSwap─────────X───────X───────iSwap────iSwap┼────────────
               │                                           │
3: ────────────┼────iSwap────────────X─────────────────────iSwap────────
               │    │
4: ────────────┼────┼────────X───X───X───X───iSwap──────────────────X───
               │    │                        │
5: ────────────┼────iSwap────X───────────X───┼──────────────────────────
               │                             │
6: ───iSwap────┼─────────────────────X───X───┼────────iSwap─────────────
      │        │                             │        │
7: ───iSwap────iSwap─────────────────────────iSwap────iSwap─────────────
              └──────────┘                           └──────────┘
```

In this example, some of the operations move forward (like the Xs on the left), and some of them move backward (like the SWAPs on the right) - but how each gate moves depends on how the gates around it move. In other words, each gate has a range of locations that it can go, and each range depends on the other operations - and oftentimes, how those other operations move matters. With that mental model, I think the only way to know for sure is to try every single combination of ranges, which is way too slow.

**Comment by mpharrigan** [comment]
but in your example, the single qubit gates haven't been merged. I'm pretty sure a strategy of:

 1. merge all single qubit gates
 2. greedily fill in alternating layers of single and two qubit moments
 3. drop any empty moments

works

**Comment by KevinVillela** [comment]
That's what I was hoping as well, but it doesn't work for the following type of circuit:
```
0: ───X───×───────────
          │
1: ───────×───────────

2: ───────×───X───×───
          │       │
3: ───────×───────×───
```
which should optimize to
```
0: ───────X───×───
              │
1: ───────────×───

2: ───×───X───×───
      │       │
3: ───×───────×───
```
I've added a test to catch this explicitly.

**Comment by maffoo** [comment]
@KevinVillela, the example circuit you showed would work with the simple algorithm outlined by @mpharrigan if step 2 runs in reverse, pushing gates toward later moments instead of earlier ones. This is generally how we do things internally because it's good to eliminate dead time after gates when qubits can decay before measurement.

**Comment by KevinVillela** [comment]
Thanks for the response @maffoo! That does work for the above example, but I don't think it will for the following:
```
0: ─X──×──X───
       │
1: ────×─────

2: ────×─────
       │       
3: ─X──×─────
```
which is already optimized, but with the algorithm of greedily grabbing the single-qubit gates from the end, would result in:
```
0: ─X──×──X───────
       │
1: ────×──────────

2: ──────────×────
             │       
3: ───────X──×────
```

**Comment by maffoo** [comment]
@KevinVillela, that's a good point; basically you need a way to decide if the last moment is going to be single qubit gates or two-qubit gates. It still seems to me that there should be a fairly simple way to decide this that keeps the spirit of what @mpharrigan proposed. Or if there's no fool-proof way to decide, the optimizer could just try both and pick the one that results in the shortest circuit.

**Comment by KevinVillela** [comment]
@maffoo ah but the fun is only just beginning! I agree that in the examples above, we just need to decide what the last moment will be, and can even try both and take the better one. But one can imagine a circuit that contains multiple combinations of each of the above circuit patterns serially and in the middle of other operations. Even if we went with an approach of subdividing the circuit somehow and deciding on a strategy, each decision affects the rest of the circuit and will affect the strategies we want to use in turn.

**Comment by maffoo** [comment]
The other assumption I am making is that we want the final circuit to consist of _alternating_ single- and two-qubit moments. I think this avoids the decision cascade you describe, once the "parity" is chosen (whether to end with a single- or two-qubit moment). We use circuits with such an alternating structure to calibrate the gates (as in the supremacy experiment) so I would expect higher fidelity with such a circuit structure. In addition, even if a particular circuit has no single-qubit gates in one of the nominal single-qubit moments, under various randomized compiling protocols we would insert gates in those slots anyway, and "fill in" the circuit while preserving the alternating moment structure. So while you're right that the general alignment problem may be quite complicated, I think having an optimizer that produces this simpler structure would be useful for near-term experiments.

**Comment by KevinVillela** [comment]
Oh okay, thanks for the info, I was not aware of that. Couple of follow-up questions:
  - I agree that the algorithm for this should be pretty straightforward, and I'm happy to write it up. That being said, do we have any evidence of it being useful in practice? I worry that the increase in fidelity is going to be undercut by an increase in circuit size and time to execute.
  - Is the current algorithm (aligning 2-qubit and single-qubit gates and minimizing circuit length instead of alternating) useful for anything? If so, I could make it into a different strategy for this optimizer, or a different optimizer entirely. If not, I'll probably just delete it.

Thanks again for the responses!

**Comment by mpharrigan** [comment]
> I worry that the increase in fidelity is going to be undercut by an increase in circuit size and time to execute

But keeping the moments homogeneous reduces the circuit duration because the two-qubit layers can be roughly half as long if they don't have to "wait" for simultaneous 1q gates that would otherwise be in that moment.

Another useful piece of context is that: after merging single qubit operations these circuits are going to be *pretty close* to having the alternating structure already. Most 2q gate compilations *need* flanking single qubit operations. People rarely want to do exactly ISWAP * CPHASE(pi/6)

**Comment by KevinVillela** [comment]
Sounds good, thanks for the comments - I'll send out another PR in the next couple of days

**Comment by KevinVillela** [comment]
Closing this in favor of https://github.com/quantumlib/Cirq/pull/2869, thanks all for the comments!

**Pull request closed without merging**